### DISCUSSION TITLE
Modifiy concurrent multiplexed responses over a single connection?

### DISCUSSION THREAD

**Discussion by RjRDRG** [abhinavsingh/proxy.py#1181] [Q&A]
Hello, i have some questions regarding the modification of HTTP 2 responses:

1. Is a new instance HttpProxyBasePlugin created for each TCP connection or is a new instance created for each request/response pair?

2. Is it possible to inspect the fields of a request in the **handle_upstream_chunk(...)** method 
(such as request .path, .method, .headers, ...) or do i need to cache the request in a instance variable? If so how can i associate a response to the the request that preceded it, do i have to use cookies?

3. Does the ModifyChunkResponsePlugin example supports the modification of concurrent multiplexed responses over a single TCP connection?

**Comment by abhinavsingh** [comment]
@RjRDRG Please find reply inline below

> 1. Is a new instance HttpProxyBasePlugin created for each TCP connection or is a new instance created for each request/response pair?

`HttpProxyBasePlugin` instance is created per request.

> 2\. Is it possible to inspect the fields of a request in the **handle_upstream_chunk(...)** method
>     (such as request .path, .method, .headers, ...) or do i need to cache the request in a instance variable? If so how can i associate a response to the the request that preceded it, do i have to use cookies?

I think if cookies works, that should be the defacto mode.   Because it will simply work out of box.

If you want to manage state within your Python plugin, you have following options:

1. Create your own multiprocessing safe data structures.   All instances of `HttpProxyBasePlugin` will have access to this data structures and cooperate based on that.
2. It really depends on your use-case.  Example, if you simply want to hold a request and wake it up based upon another incoming request, you could use Python semaphores and related utilities.
3. You could also use external resources like Sqlite, Redis etc for synchronization.  Though I'll not recommend taking this path unless you are working on cross-instance synchronizations.

Finally, there is an inbuilt `EventManager` class which can be used here with some success.  However, I personally have never used it for synchronization and mostly only for publish-subscribe patterns.

> 3\. Does the ModifyChunkResponsePlugin example supports the modification of concurrent multiplexed responses over a single TCP connection?

`ModifyChunkResponsePlugin` by default expects `HTTP/1.1` traffic.  But it can easily be modified for your use case.  Example, handle the chunk and parse it into the format that you are expecting.   Assemble the chunks and dispatch it to the client upon end of response from upstream.

Lastly, there is a connection pooling facility.  It can be enabled using `--enable-conn-pool` flag.  Internally, it will enable a connection pool for all instances of `HttpProxyBasePlugin`.  Essentially, connections gets reused when `--enable-conn-pool` is used.  You can look at its implementation and check if there is anything useful going in there.
